File: java/src/main/java/org/example/crypto/AesKeyWrapExample.java
package org.example.crypto;

import java.util.Arrays;
import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;

public final class AesKeyWrapExample {
    private AesKeyWrapExample() {}

    public static SecretKey generateAesKey() throws Exception {
        KeyGenerator generator = KeyGenerator.getInstance("AES");
        generator.init(256, CryptoSupport.RNG);
        return generator.generateKey();
    }

    public static byte[] wrap(SecretKey keyEncryptionKey, SecretKey keyToWrap) throws Exception {
        Cipher cipher = Cipher.getInstance("AESWrap");
        cipher.init(Cipher.WRAP_MODE, keyEncryptionKey);
        return cipher.wrap(keyToWrap);
    }

    public static SecretKey unwrap(SecretKey keyEncryptionKey, byte[] wrapped) throws Exception {
        Cipher cipher = Cipher.getInstance("AESWrap");
        cipher.init(Cipher.UNWRAP_MODE, keyEncryptionKey);
        return (SecretKey) cipher.unwrap(wrapped, "AES", Cipher.SECRET_KEY);
    }

    public static void main(String[] args) throws Exception {
        SecretKey kek = generateAesKey();
        SecretKey contentKey = generateAesKey();
        byte[] wrapped = wrap(kek, contentKey);
        SecretKey recovered = unwrap(kek, wrapped);
        CryptoSupport.require(Arrays.equals(contentKey.getEncoded(), recovered.getEncoded()), "AES key wrap failed");
        System.out.println("AesKeyWrapExample wrapped=" + CryptoSupport.hex(wrapped));
    }
}

File: python/aes_key_wrap_example.py
"""AES key wrap with padding for protecting key material under a KEK."""

from cryptography.hazmat.primitives.keywrap import (
    aes_key_unwrap_with_padding,
    aes_key_wrap_with_padding,
)

from crypto_support import constant_time_equal, random_bytes


def wrap(key_encryption_key: bytes, key_to_wrap: bytes) -> bytes:
    return aes_key_wrap_with_padding(key_encryption_key, key_to_wrap)


def unwrap(key_encryption_key: bytes, wrapped_key: bytes) -> bytes:
    return aes_key_unwrap_with_padding(key_encryption_key, wrapped_key)


def main() -> None:
    key_encryption_key = random_bytes(32)
    content_key = random_bytes(32)
    wrapped = wrap(key_encryption_key, content_key)
    recovered = unwrap(key_encryption_key, wrapped)
    assert constant_time_equal(content_key, recovered)
    print(f"AesKeyWrapExample wrapped={wrapped.hex()}")


if __name__ == "__main__":
    main()
